### PR TITLE
Boot delay optimizations

### PR DESCRIPTION
**Pull request by krish2718**
Improve patch download time to RPU using DMA rather than using deprecated indirect write mechanism (4 bytes at a time).

### CODE REVIEW

**Review by sr1dh48r — APPROVED** [review]

**Review by udaynordic — APPROVED** [review]

File: drivers/wifi/nrf700x/osal/hw_if/hal/src/hal_fw_patch_loader.c [hunk 1]
```diff
@@ -13,6 +13,95 @@
...
+	for (int chunk = 0; chunk < num_chunks; chunk++) {
+		unsigned char *patch_data_ram;
+		unsigned int patch_chunk_size =
+			((chunk == num_chunks - 1) ? last_chunk_size : MAX_PATCH_CHUNK_SIZE);
+		const void *src_patch_offset = (const char *)fw_patch_data +
+			chunk * MAX_PATCH_CHUNK_SIZE;
+		int dest_chunk_offset = dest_addr + chunk * MAX_PATCH_CHUNK_SIZE;
+
+		patch_data_ram = wifi_nrf_osal_mem_alloc(hal_dev_ctx->hpriv->opriv,
+									patch_chunk_size);
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/osal/hw_if/hal/src/hal_fw_patch_loader.c:49]: Can the patch_data_ram (of MAX_PATCH_CHUNK_SIZE) be allocated and freed only once instead of allocing/freeing for every chunk?

File: drivers/wifi/nrf700x/zephyr/src/zephyr_coex.c [hunk 2]
```diff
@@ -47,7 +47,7 @@ static const struct gpio_dt_spec btrf_switch_spec =
 /* PTA registers configuration of Coexistence Hardware */
 /* Separate antenna configuration, WLAN in 2.4GHz*/
 const uint16_t config_buffer_SEA[] = {
-	0x19, 0xF6, 0x8, 0x62, 0xF5,
+	0x0019, 0x00F6, 0x8, 0x62, 0xF5,
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/zephyr/src/zephyr_coex.c:50]: What is this change needed for?

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/inc/fw_B/rpu_fw_patches.h [hunk 3]
```diff
@@ -8,7 +8,7 @@
  * @brief Common structures and definations.
  */
 #ifdef CONFIG_NRF700X_RADIO_TEST
-const unsigned char wifi_nrf_umac_patch_pri_bimg[] = {
+const unsigned char __aligned(4) wifi_nrf_umac_patch_pri_bimg[] = {
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/inc/fw_B/rpu_fw_patches.h:11]: Are these changes needed any longer?

**Review by sachinthegreen — CHANGES_REQUESTED** [review]

File: drivers/wifi/nrf700x/osal/hw_if/hal/src/hal_fw_patch_loader.c (hunk 1 above)
Comment (krish2718) [reply drivers/wifi/nrf700x/osal/hw_if/hal/src/hal_fw_patch_loader.c:49]: MAX_PATCH_CHUNK_SIZE refers to chunk, so, a bit confused by the statement. But we should be doing alloc per-chunk to minimise RAM usage.

File: drivers/wifi/nrf700x/zephyr/src/zephyr_coex.c (hunk 2 above)
Comment (krish2718) [reply drivers/wifi/nrf700x/zephyr/src/zephyr_coex.c:50]: oops, unrelated, will remove.

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/inc/fw_B/rpu_fw_patches.h (hunk 3 above)
Comment (krish2718) [reply drivers/wifi/nrf700x/osal/fw_if/umac_if/inc/fw_B/rpu_fw_patches.h:11]: Not really, but this is good to have in future in case we do things differently i.e., directly use this array, but I need to use GCC attribute here instead of Zephyr define, will update.

File: drivers/wifi/nrf700x/osal/hw_if/hal/inc/pal.h [hunk 4]
```diff
@@ -47,6 +47,8 @@ static const unsigned int SOC_MMAP_ADDR_OFFSETS_MCU[] = {
 
 #endif /* RPU_CONFIG_72 */
 
+#define RPU_MCU_CORE_INDIRECT_BASE 0xC0000000
```
[the remark this answers is not in the document]
Comment (krish2718) [reply drivers/wifi/nrf700x/osal/hw_if/hal/inc/pal.h:50]: No, this is unrelated to RPU as its not an absolute RPU address, this is just a host offset, so, should be in PAL to abstract.

File: drivers/wifi/nrf700x/osal/hw_if/hal/src/hal_fw_patch_loader.c (hunk 1 above)
Comment (sachinthegreen) [reply drivers/wifi/nrf700x/osal/hw_if/hal/src/hal_fw_patch_loader.c:49]: What I meant was reusing the same allocated memory to copy the chunks rather than allocating/freeing for every chunk.

**Review by sachinthegreen — CHANGES_REQUESTED** [review]

**Review by sachinthegreen — APPROVED** [review]

File: drivers/wifi/nrf700x/osal/hw_if/hal/src/hal_fw_patch_loader.c (hunk 1 above)
Comment (krish2718) [reply drivers/wifi/nrf700x/osal/hw_if/hal/src/hal_fw_patch_loader.c:49]: Offline discussion: As allocations are fast in Zephyr, the current approach actually saves both memory and time, so ,leaving it as is.

**Review by srkanordic — APPROVED** [review]
